### DISCUSSION TITLE
Workflow: Interactively run snippets of code (breakpoint, hot reloading)

### DISCUSSION THREAD

**Discussion by Splines** [ManimCommunity/manim#3954] [Ideas]
I've just watched Grant's awesome [How I animate 3Blue1Brown](https://youtu.be/rbu7Zu5X1zI) video and was fascinated by his effective workflows where he could run part of the code and immediately see the output in an interactive window. In the video, this is between 3:32 and 6:48, with the most important part starting at 5:29.

This workflow is described [in his Readme here](https://github.com/3b1b/videos?tab=readme-ov-file#workflow) and relies on
- Manim having an interactive mode (which is different than rendering a video file and opening it).
- Sublime shortcuts that are state-aware and can interact with a terminal. E.g. Grant highlighted part of the code that had a comment on top. Apparently, the `checkpoint_paste()` method then looks for cached scene states (cached based on the comment) and renders the scene starting at that point. With the help of Sublime shortcuts, Grant is able to highlight code, press the key and immediately see the effects in an interactive window that even supports navigating the 3D scene. **All of this kind of acts as a Python debugger _breakpoint_**.

---

When I first tried Manim community a few years ago, what annoyed me was that the turnaround time is too long in my opinion, especially when working with a bit more complex scenes than the examples. Always having to render out a video file is tedious. Looking at the Manim [Getting Started guide](https://docs.manim.community/en/stable/tutorials/quickstart.html) again now, it still proposes to enter commands in the shell in order to render out _video files_ to preview a scene.

**That's why I'm asking if there is any more interactive workflow available with Manim Community?** E.g. one like the presented in the video where an interactive 3D scene is always visible in a floating window. It would be really cool if this was even available in **VSCode** as I don't use Sublime. Maybe this would be something for [`manim-sideview`](https://github.com/Rickaym/manim-sideview) @Rickaym if this kind of workflow doesn't already exist?

How do you work efficiently with the Manim Community edition? Could you share some of your tips?

**Comment by Rickaym** [comment]
Mhm, manim-sideview does have a similar workflow. It allows you to see the rendered output in realtime inside the VSCode window. The way it works is that, it is listening to vscode events and when the file is being saved or a custom shortcut is being stricken, the video is rendered and a webview is spawned in the editor. The underlying rendering command is the same command for rendering that we would usually do.

**Reply by uwezi** [reply]
> The underlying rendering command is the same command for rendering that we would usually do.

...and therefore Manim Sideview is "as slow" as it would be rendering the same scene from the command line. But are you aware that you can use `Section()`s in ManimCE.
https://docs.manim.community/en/stable/tutorials/output_and_config.html#sections

I myself don't find the rendering times problematic, even when working on quite long scenes. I sometimes break out short snippets and try some quick layout changes on these short snippets, but in general from working with Manim I have got an innate understanding and feeling on how to place objects on the scene.
I am also writing almost all my texts using LaTeX and have done so for almost 30 years now - I find most GUIs very annoying because everything is hidden in menus under menus of menus and drop-down lists...

so everyone's mileage may vary...

**Reply by Splines** [reply]
While I greatly appreciate all the efforts for Manim Community Edition (Manim CE) & Manim Sideview and what it is already doing, the workflow still feels a lot more tedious compared to Grant's ways of doing things:
- With Manim CE & Manim Sideview, the _scene is completely rendered before any output is shown_. Compare that to Grant's workflow, where the scene is **rendered "live"** and you get to see the output immediately, no waiting time (depending on processor power of course). For example, see [these few seconds of the video](https://youtu.be/rbu7Zu5X1zI?&t=365) and look at the right side of the screen.
- With Manim CE & Manim Sideview, the output is a video file (that is rendered inside a webview in VSCode for example). With Grant's code the output is as well a video stream, but one that is **interactive**. He can place his cursor inside it and rotate around in the 3D scene (which I find very useful in other software even for 2D scenes to get a better sense of the spacing). See for example [these few seconds of the video](https://youtu.be/rbu7Zu5X1zI?t=731).
- With Manim CE & Manim Sideview, I can only preview the _whole scene_. Grant's workflow is more fine-grained: he can select some lines of code of a scene that starts with a comment and then start rendering at this **particular part of the scene** (all of this triggered by a single shortcut). This is especially useful when the scene is complex and shares lots of context and where it might not be worth it to split into several classes.<br>I don't think it's an easy undertaking to get this partial scene rendering to work correctly as all the previous states of the scene have to be correctly inferred. But somehow Grant did manage to do it ;)

I was just wondering, as Manim CE is a spinoff off Grant's library, whether some similar functionality is already available for Manim CE to ease the process of creation where it's a lot of back and forth and adjusting parameters (like color, spacing, sizing etc.). Quick turnaround times would tremendously help there. In Blender I get all the interactivity, but then programming in Python is way harder over there as you don't get all the lovely functions and objects that Manim already comes packed with. On the other hand, Manim is feature-rich and eases automation and animation according to mathematical rules, but the interactivity is lacking for me. More interactivity would mean more fun during the creation since less time is spent waiting for some video to have finished rendering.

**Reply by behackl** [reply]
I would like to see, at the very least, a version of Grant's mechanism for previewing particular, short sections in a scene in the community version too, it is very handy. Our `section`-mechanism does, in principle, something similar; it is just that the interface is less agile (and that the current render loop still needs to process the entire scene, which is somewhat slow even if rendering is skipped).

It mainly is a question of resources; we have been working on a refactor of the render loop for ages, recently with some more progress on our experimental branch. But progress is slow, and people get busy with other things. Help to get things done is, of course, always welcome.

**Reply by Rickaym** [reply]
It would indeed be fantastic to see Grant's mechanism incorporated into Manim CE. 

Thanks for pointing out the Section class, @uwezi - I didn't realize it before. The speed of Sideview does seem to be somewhat unusual, and it seems to be tied to the processes of VSCode, which can occasionally be a bit cumbersome. Much optimization remains to be done there. Ideally, we'd be able to see it function at the same level as Grant's GUI.

And I'd honestly love to contribute to Manim CE at heart, but unfortunately, it is a tough effort to spare time.

**Reply by mdda** [reply]
I've made a plugin that works v similarly to Grant's workflow : `manim-interactive` (see comment below)

**Comment by VladimirFokow** [comment]
check out this new VSCode extension (it's specifically for 3b1b's manim version. Maybe it can be adapted to community?):  
https://marketplace.visualstudio.com/items?itemName=bhoov.vscode-manim

extension's repo: https://github.com/bhoov/vscode-manim

update: the repo was renamed to: https://github.com/bhoov/manim-notebook

**Reply by Splines** [reply]
Thank you so much for pointing me to this brand-new extension. Will definitely check it out and probably work with Grant's version then.

**Reply by VladimirFokow** [reply]
however I have a problem with this workflow - I don't have `checkpoint_paste` command defined at all..  
Issue: https://github.com/3b1b/videos/issues/78

**Reply by Splines** [reply]
Did you use the latest version of manim that is available on pip? I don't think it has this command available. Instead, use the latest commit on the `master` branch. Clone the repo, then do `pip install -e .` inside it, see [here](https://github.com/3b1b/manim?tab=readme-ov-file#directly).

**Reply by VladimirFokow** [reply]
ooh, thanks, will try!  
upd: It worked!

**Reply by Splines** [reply]
don't forget to uninstall the version from "online-pip" beforehand: `pip uninstall manimgl`

**Reply by Rickaym** [reply]
Interesting. I'll see if I can implement something similar in sideview for CE/GL.

**Comment by Oddball777** [comment]
Any updates on this? I'd love to put some time into this but it would help a lot if someone who knows Manim's internals better could give some insight about where to start and where the low hanging fruit are.

**Comment by mdda** [comment]
I was also a bit frustrated by the Community vs Original version thing when it came to Interactive Debugging...  So I made a 'manim-interactive' plugin (now on PyPI as https://pypi.org/project/manim-interactive/) to be used with the Community version (with the OpenGL renderer). 

It's really rough at the moment (though it works for me, using an existing VScode extension that was built for 3b1b manim) - would be happy to accept comments, PRs, etc via [the repo](https://github.com/mdda/manim-interactive)
